File: app/src/main/java/info/nightscout/androidaps/plugins/pump/medtronic/comm/history/MedtronicHistoryDecoder.java
package info.nightscout.androidaps.plugins.pump.medtronic.comm.history;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import info.nightscout.androidaps.logging.L;
import info.nightscout.androidaps.plugins.pump.common.utils.ByteUtil;
import info.nightscout.androidaps.plugins.pump.common.utils.StringUtil;
import info.nightscout.androidaps.plugins.pump.medtronic.defs.MedtronicDeviceType;
import info.nightscout.androidaps.plugins.pump.medtronic.util.MedtronicUtil;


/**
 * This file was taken from GGC - GNU Gluco Control (ggc.sourceforge.net), application for diabetes
 * management and modified/extended for AAPS.
 *
 * Author: Andy {andy.rozman@gmail.com}
 */

public abstract class MedtronicHistoryDecoder<T extends MedtronicHistoryEntry> implements MedtronicHistoryDecoderInterface<T> {

    private static final Logger LOG = LoggerFactory.getLogger(L.PUMPCOMM);

    protected ByteUtil bitUtils;

    // STATISTICS (remove at later time or not)
    protected boolean statisticsEnabled = true;
    protected Map<Integer, Integer> unknownOpCodes;
    protected Map<RecordDecodeStatus, Map<String, String>> mapStatistics;
    protected MedtronicDeviceType deviceType;


    public MedtronicHistoryDecoder() {
    }


    // public abstract <E extends MedtronicHistoryEntry> Class<E> getHistoryEntryClass();

    // public abstract RecordDecodeStatus decodeRecord(T record);

    public abstract void postProcess();


    protected abstract void runPostDecodeTasks();


    // TODO_ extend this to also use bigger pages (for now we support only 1024 pages)
    private List<Byte> checkPage(RawHistoryPage page, boolean partial) throws RuntimeException {
        List<Byte> byteList = new ArrayList<Byte>();

        // if (!partial && page.getData().length != 1024 /* page.commandType.getRecordLength() */) {
        // LOG.error("Page size is not correct. Size should be {}, but it was {} instead.", 1024,
        // page.getData().length);
        // // throw exception perhaps
        // return byteList;
        // }

        if (MedtronicUtil.getMedtronicPumpModel() == null) {
            LOG.error("Device Type is not defined.");
            return byteList;
        }

        if (page.getData().length != 1024) {
            return ByteUtil.getListFromByteArray(page.getData());
        } else if (page.isChecksumOK()) {
            return ByteUtil.getListFromByteArray(page.getOnlyData());
        } else {
            return null;
        }
    }


    public List<T> processPageAndCreateRecords(RawHistoryPage rawHistoryPage) {
        return processPageAndCreateRecords(rawHistoryPage, false);
    }


    protected void prepareStatistics() {
        if (!statisticsEnabled)
            return;

        unknownOpCodes = new HashMap<Integer, Integer>();
        mapStatistics = new HashMap<RecordDecodeStatus, Map<String, String>>();

        for (RecordDecodeStatus stat : RecordDecodeStatus.values()) {
            mapStatistics.put(stat, new HashMap<String, String>());
        }
    }


    protected void addToStatistics(MedtronicHistoryEntryInterface pumpHistoryEntry, RecordDecodeStatus status,
                                   Integer opCode) {
        if (!statisticsEnabled)
            return;

        if (opCode != null) {
            if (!unknownOpCodes.containsKey(opCode)) {
                unknownOpCodes.put(opCode, opCode);
            }
            return;
        }

        if (!mapStatistics.get(status).containsKey(pumpHistoryEntry.getEntryTypeName())) {
            mapStatistics.get(status).put(pumpHistoryEntry.getEntryTypeName(), "");
        }
    }


    protected void showStatistics() {
        StringBuilder sb = new StringBuilder();

        for (Map.Entry unknownEntry : unknownOpCodes.entrySet()) {
            StringUtil.appendToStringBuilder(sb, "" + unknownEntry.getKey(), ", ");
        }

        if (isLogEnabled())
            LOG.debug("STATISTICS OF PUMP DECODE");

        if (unknownOpCodes.size() > 0) {
            LOG.warn("Unknown Op Codes: {}", sb.toString());
        }

        for (Map.Entry<RecordDecodeStatus, Map<String, String>> entry : mapStatistics.entrySet()) {
            sb = new StringBuilder();

            if (entry.getKey() != RecordDecodeStatus.OK) {
                if (entry.getValue().size() == 0)
                    continue;

                for (Map.Entry<String, String> entrysub : entry.getValue().entrySet()) {
                    StringUtil.appendToStringBuilder(sb, entrysub.getKey(), ", ");
                }

                String spaces = StringUtils.repeat(" ", 14 - entry.getKey().name().length());

                if (isLogEnabled())
                    LOG.debug("    {}{} - {}. Elements: {}", entry.getKey().name(), spaces, entry.getValue().size(),
                            sb.toString());
            } else {
                if (isLogEnabled())
                    LOG.debug("    {}             - {}", entry.getKey().name(), entry.getValue().size());
            }
        }
    }


    private int getUnsignedByte(byte value) {
        if (value < 0)
            return value + 256;
        else
            return value;
    }


    protected int getUnsignedInt(int value) {
        if (value < 0)
            return value + 256;
        else
            return value;
    }


    public String getFormattedFloat(float value, int decimals) {
        return StringUtil.getFormatedValueUS(value, decimals);
    }


    private List<T> processPageAndCreateRecords(RawHistoryPage rawHistoryPage, boolean partial) {
        List<Byte> dataClear = checkPage(rawHistoryPage, partial);
        List<T> records = createRecords(dataClear);

        for (T record : records) {
            decodeRecord(record);
        }

        runPostDecodeTasks();

        return records;
    }

    protected boolean isLogEnabled() {
        return L.isEnabled(L.PUMPCOMM);
    }

}

File: pump/medtronic/src/main/java/info/nightscout/androidaps/plugins/pump/medtronic/comm/history/MedtronicHistoryDecoder.kt
package info.nightscout.androidaps.plugins.pump.medtronic.comm.history

import info.nightscout.androidaps.plugins.pump.medtronic.util.MedtronicUtil
import info.nightscout.pump.core.utils.ByteUtil
import info.nightscout.pump.core.utils.StringUtil
import info.nightscout.rx.logging.AAPSLogger
import info.nightscout.rx.logging.LTag
import org.apache.commons.lang3.StringUtils

/**
 * This file was taken from GGC - GNU Gluco Control (ggc.sourceforge.net), application for diabetes
 * management and modified/extended for AAPS.
 *
 *
 * Author: Andy {andy.rozman@gmail.com}
 */
abstract class MedtronicHistoryDecoder<T : MedtronicHistoryEntry?>(var aapsLogger: AAPSLogger,
                                                                   var medtronicUtil: MedtronicUtil,
                                                                   var bitUtils: ByteUtil
) : MedtronicHistoryDecoderInterface<T> {

    // STATISTICS (remove at later time or not)
    protected var statisticsEnabled = true
    protected var unknownOpCodes: MutableMap<Int, Int?> = mutableMapOf()
    protected var mapStatistics: MutableMap<RecordDecodeStatus, MutableMap<String, String>> = mutableMapOf()

    abstract fun postProcess()
    protected abstract fun runPostDecodeTasks()

    // TODO_ extend this to also use bigger pages (for now we support only 1024 pages)
    @Throws(RuntimeException::class)
    private fun checkPage(page: RawHistoryPage): MutableList<Byte> {
        if (!medtronicUtil.isModelSet) {
            aapsLogger.error(LTag.PUMPCOMM, "Device Type is not defined.")
            return mutableListOf()
        }
        return if (page.data.size != 1024) {
            page.data.toMutableList()
        } else if (page.isChecksumOK) {
            page.onlyData.toMutableList()
        } else {
            mutableListOf()
        }
    }

    fun processPageAndCreateRecords(rawHistoryPage: RawHistoryPage): MutableList<T> {
        val dataClear = checkPage(rawHistoryPage)
        val records: MutableList<T> = createRecords(dataClear)
        for (record in records) {
            decodeRecord(record)
        }
        runPostDecodeTasks()
        return records
    }

    protected fun prepareStatistics() {
        if (!statisticsEnabled) return
        // unknownOpCodes = HashMap()
        // mapStatistics = HashMap()
        for (stat in RecordDecodeStatus.values()) {
            mapStatistics[stat] = hashMapOf()
            //(mapStatistics as HashMap<RecordDecodeStatus, MutableMap<String, String>>)[stat] = hashMapOf()
        }
    }

    protected fun addToStatistics(pumpHistoryEntry: MedtronicHistoryEntryInterface, status: RecordDecodeStatus, opCode: Int?) {
        if (!statisticsEnabled) return
        if (opCode != null) {
            if (!unknownOpCodes.containsKey(opCode)) {
                unknownOpCodes[opCode] = opCode
            }
            return
        }
        if (!mapStatistics[status]!!.containsKey(pumpHistoryEntry.entryTypeName)) {
            mapStatistics[status]!!.put(pumpHistoryEntry.entryTypeName, "")
        }
    }

    protected fun showStatistics() {
        var sb = StringBuilder()
        for ((key) in unknownOpCodes) {
            StringUtil.appendToStringBuilder(sb, "" + key, ", ")
        }
        aapsLogger.info(LTag.PUMPCOMM, "STATISTICS OF PUMP DECODE")
        if (unknownOpCodes.size > 0) {
            aapsLogger.warn(LTag.PUMPCOMM, "Unknown Op Codes: $sb")
        }
        for ((key, value) in mapStatistics) {
            sb = StringBuilder()
            if (key !== RecordDecodeStatus.OK) {
                if (value.size == 0) continue
                for ((key1) in value) {
                    StringUtil.appendToStringBuilder(sb, key1, ", ")
                }
                val spaces = StringUtils.repeat(" ", 14 - key.name.length)
                aapsLogger.info(LTag.PUMPCOMM, "    ${key.name}$spaces - ${value.size}. Elements: $sb")
            } else {
                aapsLogger.info(LTag.PUMPCOMM, "    ${key.name}             - ${value.size}")
            }
        }
    }

    private fun getUnsignedByte(value: Byte): Int {
        return if (value < 0) value + 256 else value.toInt()
    }

    protected fun getUnsignedInt(value: Int): Int {
        return if (value < 0) value + 256 else value
    }

    protected fun getUnsignedInt(value: Byte): Int {
        return if (value < 0) value + 256 else value.toInt()
    }

    fun getFormattedFloat(value: Float, decimals: Int): String {
        return StringUtil.getFormattedValueUS(value, decimals)
    }

}